<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ggchord：多序列BLAST比对弦图可视化工具 • ggchord</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css"><script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="pkgdown.js"></script><meta property="og:title" content="ggchord：多序列BLAST比对弦图可视化工具"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-title-body">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ggchord</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="articles/ggchord_vignette.html">Tutorial</a>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/DangJem/ggchord/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>ggchord：多序列BLAST比对弦图可视化工具</h1>
    </div>


<p>🌐 语言切换: 【<a href="README-Hans.html">现代汉语（Han）</a> | <a href="README.html">英文（English）</a>】</p>
<div id="ggchord多序列blast比对弦图可视化工具" class="section level1">

<div class="section level2">
<h2 id="概述">概述<a class="anchor" aria-label="anchor" href="#%E6%A6%82%E8%BF%B0"></a></h2>
<p><code>ggchord</code> 是一个基于 <code>ggplot2</code> 的R函数，用于将多序列的BLAST比对结果可视化为直观的弦图，支持丰富的样式定制，可轻松展示序列间的同源区域与结构关系。<code>ggchord</code> 0.1.0版本实现了从简单多序列弦图到<strong>功能更加丰富的</strong>多序列弦图的突破性升级，能够同时展示多条序列间的比对关系： - 每条序列以圆弧或定制轨迹呈现，按比例映射长度。 - 彩色连接带（ribbon）表示序列间的比对区域，支持按相似度或来源着色。 - 配备可定制的坐标轴，精准标注序列位置与长度。 - 支持整体旋转、序列方向调整等布局优化，适配不同分析场景。</p>
<p>适用于比较基因组学、泛基因组分析、噬菌体 - 宿主序列关系等研究，帮助快速挖掘序列间的同源模式。</p>
</div>
<div class="section level2">
<h2 id="主要功能">主要功能<a class="anchor" aria-label="anchor" href="#%E4%B8%BB%E8%A6%81%E5%8A%9F%E8%83%BD"></a></h2>
<ul><li>
<strong>多序列支持</strong>：同时展示2条及以上序列的比对关系，无需局限于双序列对比。</li>
<li>
<strong>序列级定制</strong>：
<ul><li>自定义序列顺序、方向（正向/反向）、间隙与半径。</li>
<li>自动或手动指定序列颜色与标签，提升可读性。</li>
</ul></li>
<li>
<strong>精细化坐标轴</strong>：
<ul><li>每条序列配备独立坐标轴，含主/次刻度，清晰标注长度位置。</li>
<li>支持调整刻度长度、标签大小与偏移量，平衡美观与信息密度。</li>
</ul></li>
<li>
<strong>灵活的连接带样式</strong>：
<ul><li>3种配色方案（单一颜色、按查询序列、按相似度渐变）。</li>
<li>可调节连接带与序列的间隙，支持贝塞尔曲线控制点定制平滑度。</li>
</ul></li>
<li>
<strong>布局优化</strong>：整体图形可旋转，适配不同展示需求。</li>
<li>
<strong>调试模式</strong>：辅助排查数据问题，显示有效/无效比对数量。</li>
</ul></div>
<div class="section level2">
<h2 id="安装">安装<a class="anchor" aria-label="anchor" href="#%E5%AE%89%E8%A3%85"></a></h2>
<div class="section level3">
<h3 id="依赖环境">依赖环境<a class="anchor" aria-label="anchor" href="#%E4%BE%9D%E8%B5%96%E7%8E%AF%E5%A2%83"></a></h3>
<ul><li>R (≥ 3.6.0)</li>
<li>ggplot2 (≥ 3.3.0)</li>
<li>ggnewscale (≥ 0.5.0)</li>
<li>RColorBrewer</li>
</ul><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggnewscale"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RColorBrewer"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="如何安装ggchord">如何安装ggchord？<a class="anchor" aria-label="anchor" href="#%E5%A6%82%E4%BD%95%E5%AE%89%E8%A3%85ggchord"></a></h3>
<p>【审查中，暂不能用】从CRAN安装ggchord的稳定版本：</p>
<p><code>install.packages("ggchord")</code></p>
<p>如果你想要开发版本，请从GitHub安装：</p>
<p><code>devtools::install_github("DangJem/ggchord")</code> 或者 <code>install.packages("ggchord_0.2.0.tar.gz")</code></p>
</div>
</div>
<div class="section level2">
<h2 id="使用说明">使用说明<a class="anchor" aria-label="anchor" href="#%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E"></a></h2>
<div class="section level3">
<h3 id="前期数据准备">前期数据准备<a class="anchor" aria-label="anchor" href="#%E5%89%8D%E6%9C%9F%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87"></a></h3>
<p>需准备三类输入数据：</p>
<div class="section level4">
<h4 id="必须序列信息数据seq_data">【必须】序列信息数据（<code>seq_data</code>）<a class="anchor" aria-label="anchor" href="#%E5%BF%85%E9%A1%BB%E5%BA%8F%E5%88%97%E4%BF%A1%E6%81%AF%E6%95%B0%E6%8D%AEseq_data"></a></h4>
<p>TSV（制表符分隔值）文件，包含序列基本信息，必须包含以下列： - <code>seq_id</code>：序列唯一标识（如基因名、登录号） - <code>length</code>：序列长度（正数）</p>
<p>示例：</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"seq_track.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>其中，<code>seq_track.tsv</code> 文件格式如下（示例）：</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode txt"><code class="sourceCode default"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>seq_id  length</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>MT108731.1  64323</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>MT118296.1  32090</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>OQ646790.1  57367</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>OR222515.1  83080</span></code></pre></div>
<p>你可以使用以下命令从FASTA文件自动生成该表格：</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">seqkit</span> fx2tab <span class="at">-nil</span> <span class="pp">*</span>fna <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'1i seq_id\tlength'</span> <span class="op">&gt;</span> seq_track.tsv</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="可选比对数据ribbon_data">【可选】比对数据（<code>ribbon_data</code>）<a class="anchor" aria-label="anchor" href="#%E5%8F%AF%E9%80%89%E6%AF%94%E5%AF%B9%E6%95%B0%E6%8D%AEribbon_data"></a></h4>
<p>TSV（制表符分隔值）文件，包含BLAST比对结果（可通过<code>outfmt6</code>或<code>outfmt7</code>格式转换），必须包含以下列： - <code>qaccver</code>：查询序列ID（需存在于<code>seq_data$seq_id</code>） - <code>saccver</code>：目标序列ID（需存在于<code>seq_data$seq_id</code>） - <code>length</code>：比对长度 - <code>pident</code>：序列相似度（百分比） - <code>qstart</code>/<code>qend</code>：查询序列上的比对起止位置 - <code>sstart</code>/<code>send</code>：目标序列上的比对起止位置</p>
<p>可以用以下脚本对示例序列进行BLAST比对，获得outfmt7格式的结果文件：</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># 使用示例样本的fasta文件进行blast比对的脚本</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="va">seqs</span><span class="op">=</span><span class="va">(</span><span class="st">"MT108731.1"</span> <span class="st">"MT118296.1"</span> <span class="st">"OQ646790.1"</span> <span class="st">"OR222515.1"</span><span class="va">)</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="va">seqsNum</span><span class="op">=</span><span class="va">${</span><span class="op">#</span><span class="va">seqs</span><span class="op">[@]</span><span class="va">}</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="va">ext</span><span class="op">=</span><span class="st">"fna"</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="cf">for</span> <span class="kw">((</span><span class="va">i</span><span class="op">=</span><span class="dv">0</span><span class="kw">;</span> <span class="va">i</span><span class="op">&lt;</span><span class="va">seqsNum</span><span class="op">-</span><span class="dv">1</span><span class="kw">;</span> <span class="va">i</span><span class="op">++</span><span class="kw">));</span> <span class="cf">do</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="cf">for</span> <span class="kw">((</span><span class="va">j</span><span class="op">=</span><span class="va">i</span><span class="op">+</span><span class="dv">1</span><span class="kw">;</span> <span class="va">j</span><span class="op">&lt;</span><span class="va">seqsNum</span><span class="kw">;</span> <span class="va">j</span><span class="op">++</span><span class="kw">));</span> <span class="cf">do</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>    <span class="bu">echo</span> <span class="at">-e</span> <span class="st">"Running BLASTN: </span><span class="va">${seqs</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span><span class="st"> vs </span><span class="va">${seqs</span><span class="op">[</span><span class="va">$j</span><span class="op">]</span><span class="va">}</span><span class="st">"</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    <span class="ex">blastn</span> <span class="dt">\</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>      <span class="at">-outfmt</span> <span class="st">'7 qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qlen slen sstrand stitle'</span> <span class="dt">\</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>      <span class="at">-query</span> <span class="st">"</span><span class="va">${seqs</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span><span class="st">.</span><span class="va">${ext}</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>      <span class="at">-subject</span> <span class="st">"</span><span class="va">${seqs</span><span class="op">[</span><span class="va">$j</span><span class="op">]</span><span class="va">}</span><span class="st">.</span><span class="va">${ext}</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>      <span class="at">-out</span> <span class="st">"</span><span class="va">${seqs</span><span class="op">[</span><span class="va">$i</span><span class="op">]</span><span class="va">}</span><span class="st">__</span><span class="va">${seqs</span><span class="op">[</span><span class="va">$j</span><span class="op">]</span><span class="va">}</span><span class="st">.o7"</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>  <span class="cf">done</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="可选基因数据gene_data">【可选】基因数据（<code>gene_data</code>）<a class="anchor" aria-label="anchor" href="#%E5%8F%AF%E9%80%89%E5%9F%BA%E5%9B%A0%E6%95%B0%E6%8D%AEgene_data"></a></h4>
<p>TSV（制表符分隔值）文件，该数据框必须包含以下列： - <code>seq_id</code>：序列唯一标识，需与序列信息数据（<code>seq_data</code>）中的 <code>seq_id</code> 对应，例如基因名、登录号等。 - <code>start</code>：基因在序列上的起始位置。 - <code>end</code>：基因在序列上的结束位置。 - <code>strand</code>：基因所在的链方向，通常用 <code>+</code> 或 <code>-</code> 表示正向或反向。 - <code>anno</code>：基因的注释信息，如基因的功能描述。</p>
<p>示例文件 <code>gene_track.tsv</code> 的格式如下：</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode txt"><code class="sourceCode default"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>seq_id  start   end strand  anno</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>MT108731.1  100 200 +   DNA binding protein</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>MT118296.1  300 400 -   Transcription factor</span></code></pre></div>
<p>你可以通过运行 <code>gff2gene_track.R</code> 脚本，将 gff3 格式的文件转换成符合要求的基因数据表格。以下是脚本的具体内容：</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 获取当前目录下所有 gff3 文件的路径</span></span>
<span><span class="va">gff3FilesPath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"."</span>, pattern <span class="op">=</span> <span class="st">"*.gff3"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 读取所有 gff3 文件并合并成一个数据框</span></span>
<span><span class="va">gff3Table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_df</a></span><span class="op">(</span><span class="va">gff3FilesPath</span>,<span class="op">~</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">.x</span>,show_col_types <span class="op">=</span> <span class="cn">F</span>,comment <span class="op">=</span> <span class="st">"#"</span>,col_names <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rlang.r-lib.org/reference/set_names.html" class="external-link">set_names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seq_id"</span>, <span class="st">"source"</span>, <span class="st">"type"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"score"</span>, <span class="st">"strand"</span>, <span class="st">"phase"</span>, <span class="st">"attributes"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 筛选出 CDS 类型的记录，并提取注释信息</span></span>
<span><span class="va">geneTrackTable</span> <span class="op">&lt;-</span> <span class="va">gff3Table</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">type</span><span class="op">==</span><span class="st">"CDS"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>anno<span class="op">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="va">attributes</span>,<span class="st">"(?&lt;=product=)[^;]+(?=;)"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">seq_id</span>,<span class="va">start</span>,<span class="va">end</span>,<span class="va">strand</span>,<span class="va">anno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 将处理后的数据框保存为 TSV 文件</span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span><span class="va">geneTrackTable</span>,<span class="st">"gene_track.tsv"</span><span class="op">)</span></span></code></pre></div>
<p>运行上述脚本后，会在当前目录下生成 <code>gene_track.tsv</code> 文件，该文件即可作为基因数据用于后续的分析和可视化。</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="使用示例">使用示例<a class="anchor" aria-label="anchor" href="#%E4%BD%BF%E7%94%A8%E7%A4%BA%E4%BE%8B"></a></h2>
<div class="section level3">
<h3 id="数据读取">数据读取<a class="anchor" aria-label="anchor" href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96"></a></h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#读取序列长度数据</span></span>
<span><span class="va">seq_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"seq_track.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 读取并处理BLAST数据</span></span>
<span><span class="va">read_blast</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">file</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">FALSE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"qaccver"</span>,<span class="st">"saccver"</span>,<span class="st">"pident"</span>,<span class="st">"length"</span>,<span class="st">"mismatches"</span>,<span class="st">"gapopen"</span>,</span>
<span>                    <span class="st">"qstart"</span>,<span class="st">"qend"</span>,<span class="st">"sstart"</span>,<span class="st">"send"</span>,<span class="st">"evalue"</span>,<span class="st">"bitscore"</span>,</span>
<span>                    <span class="st">"qcovs"</span>,<span class="st">"qlen"</span>,<span class="st">"slen"</span>,<span class="st">"sstrand"</span>,<span class="st">"stitle"</span><span class="op">)</span></span>
<span>  <span class="va">df</span></span>
<span><span class="op">}</span></span>
<span><span class="va">blast_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"."</span>, pattern <span class="op">=</span> <span class="st">"*.o7"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">all_blast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">blast_files</span>, <span class="va">read_blast</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ribbon_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">all_blast</span>, <span class="va">length</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 读取基因注释数据，为了使图像更加美观，这里过滤掉了较短的基因注释</span></span>
<span><span class="va">gene_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"gene_track.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">end</span><span class="op">-</span><span class="va">start</span>, n <span class="op">=</span> <span class="fl">5</span>, by <span class="op">=</span> <span class="va">seq_id</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="只传入必要的seq_data">只传入必要的seq_data<a class="anchor" aria-label="anchor" href="#%E5%8F%AA%E4%BC%A0%E5%85%A5%E5%BF%85%E8%A6%81%E7%9A%84seq_data"></a></h3>
<p>对于ggchord来说，序列数据是最重要的、不可或缺的。默认情况下，序列将按你所传入的seq_data顺序，以逆时针来排列。当然这些都是可以修改的</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part1_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part1_1.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>例如，在下面的例子中，你可以通过seq_order、seq_orientation及seq_curvature来控制序列的顺序、方向和曲率，通过seq_colors来设置序列的颜色</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part1_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  seq_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MT118296.1"</span>, <span class="st">"OR222515.1"</span>, <span class="st">"MT108731.1"</span>, <span class="st">"OQ646790.1"</span><span class="op">)</span>,</span>
<span>  seq_orientation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>  seq_curvature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span>,<span class="op">-</span><span class="fl">2</span>,<span class="fl">6</span><span class="op">)</span>,</span>
<span>  seq_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steelblue"</span>, <span class="st">"orange"</span>, <span class="st">"pink"</span>, <span class="st">"yellow"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part1_2.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
</div>
<div class="section level3">
<h3 id="加入序列比对数据">加入序列比对数据<a class="anchor" aria-label="anchor" href="#%E5%8A%A0%E5%85%A5%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9%E6%95%B0%E6%8D%AE"></a></h3>
<p>对于基因比对弦图来说，序列比对绝对是我们主要关注的点，因此ribbon_data是除seq_data之外最重要的数据。 默认情况下，ribbon的填充色由blast结果中的百分比一致性决定。</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part2_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  ribbon_data <span class="op">=</span> <span class="va">ribbon_data</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part2_1.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>当然，这些也都是可以更改的。例如，你可以将填充色制定为从查询序列出发的颜色，这样用户可以更加容易捕获到不同序列间的比对</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part2_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  ribbon_data <span class="op">=</span> <span class="va">ribbon_data</span>,</span>
<span>  ribbon_color_scheme <span class="op">=</span> <span class="st">"query"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part2_2.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>如果你觉得颜色无所谓，也可以设置成单色</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part2_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  ribbon_data <span class="op">=</span> <span class="va">ribbon_data</span>,</span>
<span>  ribbon_color_scheme <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  ribbon_colors <span class="op">=</span> <span class="st">"orange"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part2_3.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>另外，ribbon会根据序列的方向、弯曲度、间距及半径等参数自动调整以完美匹配（注：坐标轴和基因箭头亦是如此） &gt; 当前版本仍存在一些问题，再某些参数组合时可能出现图像变形的情况，这将会在未来的版本中得到修复</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part2_4</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  ribbon_data <span class="op">=</span> <span class="va">ribbon_data</span>,</span>
<span>  seq_orientation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>  seq_curvature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span>,<span class="op">-</span><span class="fl">2</span>,<span class="fl">6</span><span class="op">)</span>,</span>
<span>  seq_gap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.1</span>,<span class="fl">.05</span>,<span class="fl">.09</span>,<span class="fl">.05</span><span class="op">)</span>,</span>
<span>  seq_radius <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part2_4.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="加入基因注释信息">加入基因注释信息<a class="anchor" aria-label="anchor" href="#%E5%8A%A0%E5%85%A5%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A%E4%BF%A1%E6%81%AF"></a></h2>
<p>通过基因注释信息，我们可以更加容易得找到序列的基因区域的一致性区域，以挖掘出不同物种之间的进化关系</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part3_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  gene_data <span class="op">=</span> <span class="va">gene_data</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part3_1.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>默认情况下箭头的填充色为链模式，即使用其所在基因序列的链方向进行填充，也可以使用手动指定模式，此时将使用gene_data中的anno注释类别进行颜色填充。</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part3_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  gene_data <span class="op">=</span> <span class="va">gene_data</span>,</span>
<span>  gene_color_scheme <span class="op">=</span> <span class="st">"manual"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part3_2.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>当然，基因注释标签也是可以显示的。不过，如你所间，要想调节出一个完美的效果或许会花费一番功夫</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part3_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  gene_data <span class="op">=</span> <span class="va">gene_data</span>,</span>
<span>  gene_label_show <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  gene_label_rotation <span class="op">=</span> <span class="fl">45</span>,</span>
<span>  gene_label_radial_offset <span class="op">=</span> <span class="fl">.1</span>,</span>
<span>  panel_margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>l<span class="op">=</span><span class="fl">.2</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part3_3.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
</div>
<div class="section level2">
<h2 id="综合示例">综合示例<a class="anchor" aria-label="anchor" href="#%E7%BB%BC%E5%90%88%E7%A4%BA%E4%BE%8B"></a></h2>
<p>我们一般绘图会同时使用seq_data、ribbon_data和gene_data，这样的图像才会更加赏心悦目。</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">part4_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ggchord.html">ggchord</a></span><span class="op">(</span></span>
<span>  seq_data <span class="op">=</span> <span class="va">seq_data</span>,</span>
<span>  ribbon_data <span class="op">=</span> <span class="va">ribbon_data</span>,</span>
<span>  gene_data <span class="op">=</span> <span class="va">gene_data</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part4_1.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>当然，ggchord也提供了比较丰富的参数来控制图像的各种细节。</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">part4_2</span> <span class="op">&lt;</span>- ggchord<span class="er">(</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>  <span class="ex">seq_data</span> = seq_data,</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>  <span class="ex">ribbon_data</span> = ribbon_data,</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>  <span class="ex">gene_data</span> = gene_data,</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>  <span class="ex">title</span> = <span class="st">"Multi-sequence Chord Diagram with Gene Annotations"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>  <span class="ex">seq_gap</span> = .03,</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>  <span class="ex">seq_radius</span> = c<span class="er">(</span><span class="ex">3,2,2,1</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>  <span class="ex">seq_orientation</span> = c<span class="er">(</span><span class="ex">-1,</span> <span class="at">-1,</span> <span class="at">-1,</span> 1<span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>  <span class="ex">seq_curvature</span> = c<span class="er">(</span><span class="ex">0,1,-1,1.5</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>  <span class="ex">gene_offset</span> = list<span class="er">(</span><span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=-.2</span><span class="kw">)</span><span class="ex">,</span> </span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>                     <span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=-.2</span><span class="kw">)</span><span class="ex">,</span> </span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a>                     <span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=0</span><span class="kw">)</span><span class="ex">,</span> </span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a>                     <span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=.1</span><span class="kw">))</span><span class="ex">,</span></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a>  <span class="ex">gene_label_rotation</span> = list<span class="er">(</span><span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=45,</span><span class="st">"-"</span><span class="ex">=-45</span><span class="kw">)</span><span class="ex">,</span> </span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a>                             <span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=-.2</span><span class="kw">)</span><span class="ex">,</span> </span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a>                             <span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=0</span><span class="kw">)</span><span class="ex">,</span> </span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a>                             <span class="ex">c</span><span class="er">(</span><span class="st">"+"</span><span class="ex">=.2,</span><span class="st">"-"</span><span class="ex">=.1</span><span class="kw">))</span><span class="ex">,</span></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a>  <span class="ex">gene_label_radial_offset</span> = c<span class="er">(</span><span class="ex">0,0,0,0</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a>  <span class="ex">gene_label_circum_offset</span> = c<span class="er">(</span><span class="ex">1,</span> 0, <span class="at">-2,</span> 0<span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a>  <span class="ex">gene_label_circum_limit</span> = c<span class="er">(</span><span class="ex">T,T,T,T</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-21"><a href="#cb19-21" tabindex="-1"></a>  <span class="ex">gene_width</span> = .08,</span>
<span id="cb19-22"><a href="#cb19-22" tabindex="-1"></a>  <span class="ex">gene_label_show</span> = T,</span>
<span id="cb19-23"><a href="#cb19-23" tabindex="-1"></a>  <span class="ex">gene_color_scheme</span> = <span class="st">"strand"</span>,</span>
<span id="cb19-24"><a href="#cb19-24" tabindex="-1"></a>  <span class="ex">ribbon_gap</span> = .1,</span>
<span id="cb19-25"><a href="#cb19-25" tabindex="-1"></a>  <span class="ex">ribbon_color_scheme</span> = <span class="st">"pident"</span>,</span>
<span id="cb19-26"><a href="#cb19-26" tabindex="-1"></a>  <span class="ex">ribbon_ctrl_point</span> = c<span class="er">(</span><span class="ex">0,0</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-27"><a href="#cb19-27" tabindex="-1"></a>  <span class="ex">axis_label_orientation</span> = c<span class="er">(</span><span class="ex">0,45,80,130</span><span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-28"><a href="#cb19-28" tabindex="-1"></a>  <span class="ex">axis_gap</span> = 0,</span>
<span id="cb19-29"><a href="#cb19-29" tabindex="-1"></a>  <span class="ex">axis_tick_major_number</span> = 5,</span>
<span id="cb19-30"><a href="#cb19-30" tabindex="-1"></a>  <span class="ex">axis_tick_major_length</span> = 0.03,</span>
<span id="cb19-31"><a href="#cb19-31" tabindex="-1"></a>  <span class="ex">axis_tick_minor_number</span> = 5,</span>
<span id="cb19-32"><a href="#cb19-32" tabindex="-1"></a>  <span class="ex">axis_tick_minor_length</span> = 0.01,</span>
<span id="cb19-33"><a href="#cb19-33" tabindex="-1"></a>  <span class="ex">axis_label_size</span> = 2,</span>
<span id="cb19-34"><a href="#cb19-34" tabindex="-1"></a>  <span class="ex">axis_label_offset</span> = 2,</span>
<span id="cb19-35"><a href="#cb19-35" tabindex="-1"></a>  <span class="ex">rotation</span> = 45, </span>
<span id="cb19-36"><a href="#cb19-36" tabindex="-1"></a>  <span class="ex">show_axis</span> = T,</span>
<span id="cb19-37"><a href="#cb19-37" tabindex="-1"></a>  <span class="ex">panel_margin</span> = list<span class="er">(</span><span class="va">t</span><span class="op">=</span>.1<span class="kw">)</span><span class="ex">,</span></span>
<span id="cb19-38"><a href="#cb19-38" tabindex="-1"></a>  <span class="ex">debug</span> = TRUE,</span>
<span id="cb19-39"><a href="#cb19-39" tabindex="-1"></a><span class="kw">)</span></span></code></pre></div>
<div class="float">
<img src="/examples/plots/v1.0.0/part4_2.jpg" alt="plot"><div class="figcaption">plot</div>
</div>
<p>不过，当前还是软件的早期版本，很多地方还有不完善的地方甚至是BUG。随着日后软件迭代，相信这些问题都将一一解决。</p>
</div>
<div class="section level2">
<h2 id="参数详情">参数详情<a class="anchor" aria-label="anchor" href="#%E5%8F%82%E6%95%B0%E8%AF%A6%E6%83%85"></a></h2>
<table class="table"><colgroup><col width="20%"><col width="20%"><col width="20%"><col width="20%"><col width="20%"></colgroup><thead><tr><th>参数类别</th>
<th>参数名</th>
<th>类型</th>
<th>默认值</th>
<th>描述</th>
</tr></thead><tbody><tr><td><strong>核心数据</strong></td>
<td><code>seq_data</code></td>
<td>data.frame/tibble</td>
<td>-</td>
<td>包含序列信息的数据框，必须包含列：<br> - <code>seq_id</code>：序列唯一标识<br> - <code>length</code>：序列长度</td>
</tr><tr><td></td>
<td><code>ribbon_data</code></td>
<td>data.frame/tibble</td>
<td>-</td>
<td>包含BLAST比对结果的数据框，可选。若提供，则必须包含列：<br> - <code>qaccver</code>：查询序列ID<br> - <code>saccver</code>：目标序列ID<br> - <code>length</code>：比对长度<br> - <code>pident</code>：序列相似度百分比<br> - <code>qstart</code>：查询序列起始位置<br> - <code>qend</code>：查询序列结束位置<br> - <code>sstart</code>：目标序列起始位置<br> - <code>send</code>：目标序列结束位置</td>
</tr><tr><td></td>
<td><code>gene_data</code></td>
<td>data.frame/tibble</td>
<td>-</td>
<td>包含基因注释信息的数据框，可选。若提供，则必须包含列：<br> - <code>seq_id</code>：序列唯一标识<br> - <code>start</code>：基因起始位置<br> - <code>end</code>：基因结束位置<br> - <code>strand</code>：链方向（+或-）<br> - <code>anno</code>：基因注释</td>
</tr><tr><td><strong>基本样式</strong></td>
<td><code>title</code></td>
<td>字符串</td>
<td>NULL</td>
<td>图形主标题</td>
</tr><tr><td><strong>序列布局</strong></td>
<td><code>seq_order</code></td>
<td>字符向量</td>
<td>NULL</td>
<td>指定序列绘制顺序；若为NULL则使用 <code>seq_data</code> 中的顺序</td>
</tr><tr><td></td>
<td><code>seq_labels</code></td>
<td>字符向量或命名向量</td>
<td>NULL</td>
<td>序列标签；若为NULL则使用 <code>seq_id</code>
</td>
</tr><tr><td></td>
<td><code>seq_orientation</code></td>
<td>数值向量或单值</td>
<td>1</td>
<td>每条序列的方向：1（正向）或 -1（反向）；默认正向</td>
</tr><tr><td></td>
<td><code>seq_gap</code></td>
<td>数值或向量</td>
<td>0.03</td>
<td>长度与序列数一致，定义每条序列头部到下一条序列尾部的弧度比例 [0,0.5)</td>
</tr><tr><td></td>
<td><code>seq_radius</code></td>
<td>数值或向量</td>
<td>1.0</td>
<td>序列圆弧半径，支持单值或与序列数相同的向量</td>
</tr><tr><td></td>
<td><code>seq_curvature</code></td>
<td>数值或向量</td>
<td>1.0</td>
<td>序列圆弧的弯曲程度：1为标准圆弧，0为直线，&gt;1更弯曲</td>
</tr><tr><td></td>
<td><code>seq_colors</code></td>
<td>颜色向量或命名向量</td>
<td>NULL</td>
<td>定义各序列圆弧颜色；若为NULL则基于 RColorBrewer Set1 自动生成</td>
</tr><tr><td><strong>基因样式</strong></td>
<td><code>gene_offset</code></td>
<td>数值、向量或列表</td>
<td>0.03</td>
<td>基因箭头与序列圆弧之间的径向偏移距离。支持：<br> - 单值：所有序列的所有链使用相同偏移<br> - 向量：长度与序列数一致，每个序列的所有链使用相同偏移<br> - 列表：命名列表，每个元素对应一个序列，元素可为单值（该序列所有链）或包含”+“和”-“的命名向量（区分链）</td>
</tr><tr><td></td>
<td><code>gene_width</code></td>
<td>数值或向量</td>
<td>0.1</td>
<td>基因箭头宽度</td>
</tr><tr><td></td>
<td><code>gene_label_show</code></td>
<td>逻辑值</td>
<td>FALSE</td>
<td>是否显示基因标签</td>
</tr><tr><td></td>
<td><code>gene_label_rotation</code></td>
<td>数值、向量或列表</td>
<td>0</td>
<td>基因标签的旋转角度（度），支持与 <code>gene_offset</code> 相同的参数格式</td>
</tr><tr><td></td>
<td><code>gene_label_size</code></td>
<td>数值</td>
<td>2.5</td>
<td>基因注释文字大小</td>
</tr><tr><td></td>
<td><code>gene_label_radial_offset</code></td>
<td>数值、向量或列表</td>
<td>0</td>
<td>基因标签相对于箭头的径向偏移（正值向外，负值向内），支持与 <code>gene_offset</code> 相同的参数格式</td>
</tr><tr><td></td>
<td><code>gene_label_circum_offset</code></td>
<td>数值、向量或列表</td>
<td>0</td>
<td>基因标签沿序列周向的偏移比例（相对于基因长度），支持与 <code>gene_offset</code> 相同的参数格式</td>
</tr><tr><td></td>
<td><code>gene_label_circum_limit</code></td>
<td>逻辑值、向量或列表</td>
<td>TRUE</td>
<td>是否限制周向偏移不超过基因长度的一半，支持与 <code>gene_offset</code> 相同的参数格式</td>
</tr><tr><td></td>
<td><code>gene_color_scheme</code></td>
<td>字符</td>
<td>“strand”</td>
<td>指定基因颜色方案，可选”strand”（按链方向）或”manual”（手动指定）</td>
</tr><tr><td></td>
<td><code>gene_colors</code></td>
<td>颜色向量</td>
<td>-</td>
<td>用于指定基因箭头的填充色，具体行为取决于 <code>gene_color_scheme</code>：<br> - “strand”模式：支持命名向量（仅”+“/”-“）、非命名向量（先”+“后”-“）或单值（正负链同色），缺省则”+“为红色、”-“为蓝色<br> -”manual”模式：支持命名向量（对应 <code>anno</code>）、非命名向量（截断多余，补齐不足），缺省则使用自动生成颜色</td>
</tr><tr><td></td>
<td><code>gene_order</code></td>
<td>字符向量</td>
<td>NULL</td>
<td>指定基因在图例中的显示顺序；若为NULL则使用基因在数据中出现的顺序</td>
</tr><tr><td><strong>连接带样式</strong></td>
<td><code>ribbon_color_scheme</code></td>
<td>字符</td>
<td>“pident”</td>
<td>连接带配色方案，可选 “single”、“query” 或 “pident”</td>
</tr><tr><td></td>
<td><code>ribbon_colors</code></td>
<td>-</td>
<td>-</td>
<td>连接带颜色参数：<br> - single：单一颜色（单值或向量取第一个）<br> - query：按查询序列映射颜色（命名或非命名向量或单值）<br> - pident：渐变色阶向量，用于按相似度百分比生成渐变，默认使用蓝到黄的渐变色</td>
</tr><tr><td></td>
<td><code>ribbon_alpha</code></td>
<td>数值</td>
<td>0.35</td>
<td>连接带透明度 [0,1]</td>
</tr><tr><td></td>
<td><code>ribbon_ctrl_point</code></td>
<td>向量或列表</td>
<td>-</td>
<td>贝塞尔控制点，用于调整连接带形状：<br> - 向量：长度为2（单控制点）或4（c1x,c1y,c2x,c2y，双控制点）<br> - 列表：每个元素为包含1 - 2个控制点的子列表，默认自动计算</td>
</tr><tr><td></td>
<td><code>ribbon_gap</code></td>
<td>数值或向量</td>
<td>0.15</td>
<td>序列圆弧与连接带之间的径向距离</td>
</tr><tr><td><strong>坐标轴设置</strong></td>
<td><code>axis_gap</code></td>
<td>数值或向量</td>
<td>0.04</td>
<td>坐标轴与序列圆弧之间径向距离，支持负值</td>
</tr><tr><td></td>
<td><code>axis_tick_major_number</code></td>
<td>整数或向量</td>
<td>5</td>
<td>每条序列主刻度数</td>
</tr><tr><td></td>
<td><code>axis_tick_major_length</code></td>
<td>数值或向量</td>
<td>0.02</td>
<td>主刻度线长度比例</td>
</tr><tr><td></td>
<td><code>axis_tick_minor_number</code></td>
<td>整数或向量</td>
<td>4</td>
<td>每两个主刻度之间的次刻度数</td>
</tr><tr><td></td>
<td><code>axis_tick_minor_length</code></td>
<td>数值或向量</td>
<td>0.01</td>
<td>次刻度线长度比例</td>
</tr><tr><td></td>
<td><code>axis_label_size</code></td>
<td>数值或向量</td>
<td>3</td>
<td>坐标轴刻度文字大小</td>
</tr><tr><td></td>
<td><code>axis_label_offset</code></td>
<td>数值或向量</td>
<td>0</td>
<td>坐标轴标签相对于刻度线的偏移比例</td>
</tr><tr><td></td>
<td><code>axis_label_orientation</code></td>
<td>字符、数值或向量</td>
<td>“horizontal”</td>
<td>坐标轴标签方向：<br> - “horizontal”：水平方向<br> - 数值：旋转角度（度）<br> - 向量：长度与序列数一致或命名向量</td>
</tr><tr><td><strong>布局设置</strong></td>
<td><code>rotation</code></td>
<td>数值</td>
<td>45</td>
<td>整个图形的旋转角度（度）</td>
</tr><tr><td></td>
<td><code>panel_margin</code></td>
<td>列表</td>
<td>list(t=0,r=0,b=0,l=0)</td>
<td>图形边缘留白（t=上, r=右, b=下, l=左）</td>
</tr><tr><td><strong>图例与调试</strong></td>
<td><code>show_legend</code></td>
<td>逻辑值</td>
<td>TRUE</td>
<td>是否显示图例</td>
</tr><tr><td></td>
<td><code>show_axis</code></td>
<td>逻辑值</td>
<td>TRUE</td>
<td>是否显示坐标轴及刻度</td>
</tr><tr><td></td>
<td><code>debug</code></td>
<td>逻辑值</td>
<td>FALSE</td>
<td>是否输出调试信息</td>
</tr></tbody></table></div>
<div class="section level2">
<h2 id="图形解读">图形解读<a class="anchor" aria-label="anchor" href="#%E5%9B%BE%E5%BD%A2%E8%A7%A3%E8%AF%BB"></a></h2>
<ul><li>
<strong>序列圆弧</strong>：每条彩色圆弧代表一条序列，长度按比例映射，箭头指示方向（正向/反向）。</li>
<li>
<strong>连接带</strong>：连接不同序列的彩色区域，代表比对区间：
<ul><li>按相似度着色时，颜色渐变反映序列一致性（如从蓝色到红色表示相似度从低到高）。</li>
<li>按查询序列着色时，同一颜色的连接带来自同一条查询序列。</li>
</ul></li>
<li>
<strong>坐标轴</strong>：每条序列外侧的刻度线与数字，标注序列位置（单位与序列长度一致），方便定位比对区域。</li>
</ul></div>
<div class="section level2">
<h2 id="版本更新记录">版本更新记录<a class="anchor" aria-label="anchor" href="#%E7%89%88%E6%9C%AC%E6%9B%B4%E6%96%B0%E8%AE%B0%E5%BD%95"></a></h2>
<div class="section level3">
<h3 id="v020最新">v0.2.0（最新）<a class="anchor" aria-label="anchor" href="#v020%E6%9C%80%E6%96%B0"></a></h3>
<ul><li>
<strong>高级弧线和直线模式优化</strong>：
<ul><li>通过增强的曲线拟合算法，在双序列可视化中，“弧线模式”和“直线模式”在不同比对区域之间实现了更平滑的过渡。例如，在可视化两条密切相关且具有多个短比对的序列时，弧线或直线现在能够更优雅地连接这些区域，减少视觉混乱。</li>
<li>确保了比对数据的更准确表示。在以前的版本中，尤其是对于长距离比对，可视化可能存在轻微的失真。现在，弧线和直线的长度和位置更精确地匹配实际比对坐标，提供了更真实的序列关系视图。</li>
</ul></li>
<li>
<strong>精确的曲率和间隙控制</strong>：
<ul><li>用户现在可以更精确地控制弦图中弧线的曲率。引入了一个新参数，允许逐步调整弧线曲率，使用户能够突出不同类型的比对模式。例如，可以使用更明显的曲率来强调高度保守的区域，而较平坦的曲线则用于不太重要的比对。</li>
<li>序列和连接带之间的间隙可以进行更精细的调整。这对于可视化复杂的比对场景非常有用，其中不同序列具有不同程度的相似性。用户现在可以为不同的序列对设置不同的间隙值，确保可视化既清晰又有信息量。</li>
</ul></li>
<li>
<strong>增强的颜色定制</strong>：
<ul><li>0.2.0版本为序列和连接带提供了更广泛的调色板。除了现有的默认配色方案外，用户现在可以从各种预定义的调色板中进行选择，这些调色板针对不同类型的数据可视化进行了优化。例如，有专门设计用于突出高对比度区域的调色板，以及用于创建更微妙和美观可视化的调色板。</li>
<li>在使用<code>pident</code>配色方案为连接带着色时，用户现在可以自定义渐变颜色。这允许进行更个性化的可视化，更好地表示序列相似性的分布。例如，用户可以选择使用类似热图的渐变来清晰显示从低到高的相似性值范围。 ### v0.1.0</li>
</ul></li>
<li>支持通过<code>seq_data</code>、<code>ribbon_data</code>与<code>gene_data</code>分离管理序列、比对于基因数据。</li>
<li>新增序列方向控制、自定义顺序、间隙与半径调整。</li>
<li>实现可定制坐标轴（主/次刻度、标签位置）。</li>
<li>连接带支持3种配色方案（单一色、按查询序列、按相似度渐变）。</li>
<li>支持整体图形旋转与调试模式。 ### v0.0.2</li>
<li>支持多序列展示（从双序列升级）</li>
<li>新增双序列的“弧线模式”与“直线模式”切换。</li>
<li>支持弧线曲率调整与间隙控制。 ### v0.0.1</li>
<li>初始版本，支持双序列BLAST比对弦图可视化。</li>
</ul></div>
</div>
<div class="section level2">
<h2 id="贡献与反馈">贡献与反馈<a class="anchor" aria-label="anchor" href="#%E8%B4%A1%E7%8C%AE%E4%B8%8E%E5%8F%8D%E9%A6%88"></a></h2>
<p>欢迎提交issue报告bug或建议新功能，也可通过Pull Request参与开发。使用中遇到问题，可参考示例代码或开启调试模式（<code>debug = TRUE</code>）排查数据问题。</p>
</div>
<div class="section level2">
<h2 id="许可证">许可证<a class="anchor" aria-label="anchor" href="#%E8%AE%B8%E5%8F%AF%E8%AF%81"></a></h2>
<p>本项目采用MIT许可证，详情见<a href="LICENSE">LICENSE</a>文件。</p>
</div>
</div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Jem Dang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

